app/scripts/metamask-controller.js

```diff
@@ -3494,43 +3494,32 @@ export default class MetamaskController extends EventEmitter {
         this._convertMnemonicToWordlistIndices(seedPhraseAsBuffer),
       );
 
-      // Scan accounts until we find an empty one
-      const { chainId } = this.networkController.state.providerConfig;
+      // Scan accounts and apply checks
       const ethQuery = new EthQuery(this.provider);
       const accounts = await this.keyringController.getAccounts();
-      let address = accounts[accounts.length - 1];
-
-      for (let count = accounts.length; ; count++) {
-        const balance = await this.getBalance(address, ethQuery);
-
-        if (balance === '0x0') {
-          // This account has no balance, so check for tokens
-          await this.detectTokensController.detectNewTokens({
-            selectedAddress: address,
-          });
-
-          const tokens =
-            this.tokensController.state.allTokens?.[chainId]?.[address];
-          const detectedTokens =
-            this.tokensController.state.allDetectedTokens?.[chainId]?.[address];
-
-          if (
-            (tokens?.length ?? 0) === 0 &&
-            (detectedTokens?.length ?? 0) === 0
-          ) {
-            // This account has no balance or tokens
-            if (count !== 1) {
-              await this.removeAccount(address);
-            }
-            break;
-          }
+      let count = accounts.length;
```

Comment:
Can all this new work be done concurrently (`await Promise.all(...)`) instead of awaiting for each iteration?

Reply:
I think `keyringController.addNewAccount(i)` would need to be called a bunch of times up front to generate the addresses.  Since it must be called in i, i+1, .. order.  But once you have the addresses, they could be queried in parallel.